### DISCUSSION TITLE
Can Sofar Solar settings be set?

### DISCUSSION THREAD

**Discussion by rowra44** [davidrapan/ha-solarman#331] [Q&A]
Hey, I'd like to be able to change parameters other than change Power storage mode. For example the parameters that belong to the power storage modes other than the simple "self use mode". But maybe also battery DOD etc. My inverter type is Sofar HYD 5KTL-3PH and I'm using G3YD yaml. It's working but the only setting I have is power storage mode. Is there any way? I can also help debug, test or even help code things, just don't really know where to start hence I'm asking you first

thanks!

**Comment by davidrapan** [comment]
Hi @markv9401, did you look for settings (you would like to have access to) in the documentation? You need their register address and know the format and how are the values encoded (usually just simple integer).

**Reply by rowra44** [reply]
Yes. I'm not actually very sure what I want to control yet as I'm new to solar systems generally. But I have the whole modbus specs/docs. It's a Sofar Solar HYD 5KTL-3PH with a Solarman LSW-3 USB stick. When I downloaded the newest FW for the inverter, it also included the modbus specs. and address description. It's for the G3, which is what I'm using with ha-solarman (g3hyd.yml). This one is V1.22, though. Anyway, I think anything that's worth looking into for automations and such is in the XLSX / Address Description sheet / 4.1 Basic configuration (0x1000-0x10FF) and maybe 4.2 Remote control (0x1100-0x12FF).

Also, writing registers is 0x13 according to Sofar, in their implementation, so the services don't work for writing


[SOFAR_Modbus_Protocol_English_G3_V1.22(only 3PH).xlsx](https://github.com/user-attachments/files/18219306/SOFAR_Modbus_Protocol_English_G3_V1.22.only.3PH.xlsx)

**Reply by davidrapan** [reply]
It's usually better to add configurable entities "by their usefulness" so as soon as you will have something to control in mind let me know. 😉

_BTW, thanks for the docs._

**Reply by rowra44** [reply]
There are two I'm thinking about right now:

- IV Curve Scan
If I understand correctly, you'd have to write 2 registers at once. Start at 0x1025 and length of 2. The actual period is at 0x1026, minutes, U16 type, steps of 1,   5 < x < 65535
- The whole of Timed charging and discharging rulemaking from 0x1111 to 0x111F

I wonder (as a novice with solar systems like I said) if it's even worth trying to write rules better than the inverter itself would do otherwise, though


Also another question (let me know if I should open another "ticket" for this as this is different): I captured some traffic from the LSW-3 going to the cloud. Unsurprisingly, it is solarman v5. In the hidden menu (config_hide.html) I can rewrite all IPs/domains to one of my own thus redirecting all traffic from the cloud to HA, for example. Would this approach be better or worse than polling actively? What I propose is whether it's better to keep your integration as a "client" rather than changing it into a "server" (or both, really).

**Reply by davidrapan** [reply]
First one, could be bit tricky (as they are two different things spanned across multiple registers) but I'll look at it and second one is no brainer. 😉 I 'll let you know whenever there is anything to test out.

> I wonder (as a novice with solar systems like I said) if it's even worth trying to write rules better than the inverter itself would do otherwise, though

That depends, I'm for example using TOU/Timed for charging during off peak hours w/ active managing of morning charging window according to solar forecast (from solcast) and it saves quite some 💲. But it really depends on your specific case (prices, battery capacity, etc...)

> Would this approach be better or worse than polling actively?

This is not yet supported by the integration (but I'd like that in the future for sure!). That being said It would be rather used in combination w/ polling as that's the only way to have really "now" data about voltage, current, power, etc.

BUT, if you do not need solarman app/deye cloud (read data in their app from their servers) you can clear out (or redirect to some "blachhole") their ip from that config_hide page and you will **drastically** improve communication performance between HA and inverter: NO TIMEOUTS AT ALL. 😉

**Comment by davidrapan** [comment]
You can try Timed Programs controls - [feat: Add Timed Programs control settings - Sofar G3](https://github.com/davidrapan/ha-solarman/commit/75a2ca59dcbf5d5883b9638421b90d604eebce94)

**Reply by rmfujitsu** [reply]
Hi. i tried this fix for the sofar HYD. There is a typo in Line 1405. (1 digit to much in the register). (sorry, not a full fledged dev, so i dont mark it on your code directly)

**Reply by rmfujitsu** [reply]
and a sidenote, if a new discussion for this is needed let me know: when setting the storage mode to self use i get timeouts when using an automation to trigger it. seems like runtimes above 30 seconds go to timeout. With my current update i get update intervals around 70s most of the time. are the 30seconds set somewhere in solarman?

**Reply by davidrapan** [reply]
Can you please record debug log during this behavior?

**Reply by rmfujitsu** [reply]
there you go
[home-assistant_solarman_2025-01-20T10-36-12.935Z.log](https://github.com/user-attachments/files/18476071/home-assistant_solarman_2025-01-20T10-36-12.935Z.log)
![image](https://github.com/user-attachments/assets/e83aecc7-1ed8-463d-b1d8-f0d7766a1a93)
![image](https://github.com/user-attachments/assets/4e13794f-2037-4d5d-ad30-718088551d41)

**Reply by davidrapan** [reply]
I see you have Ethernet Logger. Which fw version are you running on the stick? There are some which are causing timeouts and it's solvable by fw upgrade of the stick. 😉

Edit: You can read more 'bout it in #326

**Reply by rmfujitsu** [reply]
yes, im on LSE3 ethernet stick. seems like i need to downgrade to 1.09.
thx!

**Reply by davidrapan** [reply]
Or upgrade to 5.09, it depends on hw type of your stick but solarman support will tell you! 😉

**Reply by rmfujitsu** [reply]
i guess that 5.09 is a typo. never saw anything anywhere about a 5.09 firmware for the lse3.

**Reply by davidrapan** [reply]
It's not a typo! This version is just for different hardware version of the stick.

**Reply by rmfujitsu** [reply]
alright. thanks for clearing that up. looking forward to the answer from solarman.
